core/core.js

```diff
@@ -1,411 +1,123 @@
+      opAsyncHandler(arguments[i], arguments[i + 1]);
     }
-    setMeta(index, end, opId);
-    assert(alignedEnd % 4 === 0);
-    assert(end - off == buf.byteLength);
-    sharedBytes.set(buf, off);
-    shared32[INDEX_NUM_RECORDS] += 1;
-    shared32[INDEX_HEAD] = alignedEnd;
-    return true;
   }
 
-  /// Returns null if empty.
-  function shift() {
-    const i = shared32[INDEX_NUM_SHIFTED_OFF];
-    if (size() == 0) {
-      assert(i == 0);
-      return null;
-    }
-
-    const off = getOffset(i);
-    const [opId, end] = getMeta(i);
-
-    if (size() > 1) {
-      shared32[INDEX_NUM_SHIFTED_OFF] += 1;
-    } else {
-      reset();
-    }
-
-    assert(off != null);
-    assert(end != null);
-    const buf = sharedBytes.subarray(off, end);
-    return [opId, buf];
+  function dispatch(opName, promiseId, control, zeroCopy) {
+    return send(opsCache[opName], promiseId, control, zeroCopy);
   }
 
-  ////////////////////////////////////////////////////////////////////////////////////////////
-  ////////////////////////////////////// Error handling //////////////////////////////////////
-  ////////////////////////////////////////////////////////////////////////////////////////////
-
-  const errorMap = {};
-
   function registerErrorClass(errorName, className, args) {
     if (typeof errorMap[errorName] !== "undefined") {
       throw new TypeError(`Error class for "${errorName}" already registered`);
     }
     errorMap[errorName] = [className, args ?? []];
   }
 
-  function handleError(className, message) {
-    if (typeof errorMap[className] === "undefined") {
-      return new Error(
-        `Unregistered error class: "${className}"\n` +
-          `  ${message}\n` +
-          `  Classes of errors returned from ops should be registered via Deno.core.registerErrorClass().`,
-      );
-    }
-
-    const [ErrorClass, args] = errorMap[className];
-    return new ErrorClass(message, ...args);
-  }
-
-  ////////////////////////////////////////////////////////////////////////////////////////////
-  ////////////////////////////////////// Async handling //////////////////////////////////////
-  ////////////////////////////////////////////////////////////////////////////////////////////
-
-  let asyncHandlers = [];
-
-  function setAsyncHandler(opId, cb) {
-    assert(opId != null);
-    asyncHandlers[opId] = cb;
+  function getErrorClassAndArgs(errorName) {
+    return errorMap[errorName] ?? [undefined, []];
   }
 
-  function handleAsyncMsgFromRust() {
-    while (true) {
-      const opIdBuf = shift();
-      if (opIdBuf == null) {
-        break;
-      }
-      assert(asyncHandlers[opIdBuf[0]] != null);
-      asyncHandlers[opIdBuf[0]](opIdBuf[1], true);
-    }
+  let nextPromiseId = 1;
+  const promiseTable = new Map();
 
-    for (let i = 0; i < arguments.length; i += 2) {
-      asyncHandlers[arguments[i]](arguments[i + 1], false);
+  function processResponse(res) {
+    // const [ok, err] = res;
+    if (res[1] === null) {
```

Comment:
Is it guaranteed that `err` will be null? Maybe use non-strict comparison instead?

Reply:
It is guaranteed to be `null` since it's the output of `to_v8()` on a `None` Option, which will always be `null`.

But the method of returning errors will likely be changed by one of my follow-up PRs to improve the performance